File: 백준/Gold/9663. N－Queen/N－Queen.py
cnt = 0
def n_queen(i):
  global cnt
  if i == N:
    cnt += 1
    return
  for j in range(N):
    if not col[j] and not diag1[i + j] and not diag2[i - j + N - 1]:
      col[j] = True
      diag1[i + j] = True
      diag2[i - j + N - 1] = True
      n_queen(i + 1)
      col[j] = False
      diag1[i + j] = False
      diag2[i - j + N - 1] = False
    
N = int(input())
result = []
# 해당 열에 퀸을 둘 수 있는지 여부
col = [False] * N
# 해당 대각선(╱)에 퀸을 둘 수 있는지 여부
diag1 = [False] * (2 * N - 1)
# 해당 대각선(╲)에 퀸을 둘 수 있는지 여부
diag2 = [False] * (2 * N - 1)
n_queen(0)
print(cnt)

File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.io.*;
import java.util.*;

public class Main {
    static int N, cnt;
    static boolean[] row, col, diag1, diag2;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        row = new boolean[N];
        col = new boolean[N];
        diag1 = new boolean[2 * N - 1];
        diag2 = new boolean[2 * N - 1];
        backtracking(0);
        System.out.print(cnt);
    }
    static void backtracking(int depth) {
        if (depth == N) {
            cnt++;
            return;
        }
        for (int i = 0; i < N; i++) {
            if (col[i]) continue;
            if (diag1[depth - i + N - 1]) continue;
            if (diag2[depth + i]) continue;
            col[i] = true;
            diag1[depth - i + N - 1] = true;
            diag2[depth + i] = true;
            backtracking(depth + 1);
            col[i] = false;
            diag1[depth - i + N - 1] = false;
            diag2[depth + i] = false;
        }
    }
}
